File: src/main/java/com/pauljoda/modularsystems/storage/tiles/TileStorageRemote.java
package com.pauljoda.modularsystems.storage.tiles;

import com.pauljoda.modularsystems.storage.container.ContainerStorageRemote;
import com.pauljoda.modularsystems.storage.gui.GuiStorageRemote;
import com.pauljoda.modularsystems.storage.items.ItemStorageRemote;
import com.teambr.bookshelf.collections.InventoryTile;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.world.World;

/**
 * Modular-Systems
 * Created by Dyonovan on 01/08/15
 */
public class TileStorageRemote extends TileStorageBasic {
    protected InventoryTile inventory;

    public TileStorageRemote() {
        inventory = new InventoryTile(2);
    }

    @Override
    public void addedToNetwork() {
        worldObj.setBlockMetadataWithNotify(xCoord, yCoord, zCoord, 0, 2);
    }

    @Override
    public void removedFromNetwork() {
        worldObj.setBlockMetadataWithNotify(xCoord, yCoord, zCoord, 5, 2);
    }

    /*******************************************************************************************************************
     ***************************************** IInventory Methods ******************************************************
     *******************************************************************************************************************/
    @Override
    public int getSizeInventory() {
        return inventory.getSizeInventory();
    }

    @Override
    public ItemStack getStackInSlot(int slot) {
        return inventory.getStackInSlot(slot);
    }

    @Override
    public ItemStack decrStackSize(int slot, int count) {
        ItemStack itemStack = inventory.getStackInSlot(slot);
        if (itemStack != null) {
            if (itemStack.stackSize <= count) {
                setInventorySlotContents(slot, null);
            }
            itemStack = itemStack.splitStack(count);
        }
        worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
        return itemStack;
    }

    @Override
    public ItemStack getStackInSlotOnClosing(int slot) {
        ItemStack stack = getStackInSlot(slot);
        if (stack != null) {
            setInventorySlotContents(slot, null);
        }
        return stack;
    }

    @Override
    public void setInventorySlotContents(int slot, ItemStack stack) {
        inventory.setStackInSlot(stack, slot);
    }

    /**
     * Returns the name of the inventory
     */
    public String getInventoryName() {
        return null;
    }

    /**
     * Returns if the inventory is named
     */
    public boolean hasCustomInventoryName() {
        return false;
    }

    @Override
    public int getInventoryStackLimit() {
        return 1;
    }

    /**
     * Do not give this method the name canInteractWith because it clashes with Container
     */
    public boolean isUseableByPlayer(EntityPlayer player) {
        return true;
    }

    @Override
    public void openInventory() {
    }

    @Override
    public void closeInventory() {
    }

    @Override
    public boolean isItemValidForSlot(int slot, ItemStack stack) {
        return slot == 0 && stack.getItem() instanceof ItemStorageRemote;
    }

    /*******************************************************************************************************************
     *********************************************** Tile Methods ******************************************************
     *******************************************************************************************************************/

    @Override
    public void updateEntity() {
        super.updateEntity();

        if (getCore() != null && inventory.getStackInSlot(0) != null && inventory.getStackInSlot(1) == null &&
                inventory.getStackInSlot(0).getItem() instanceof ItemStorageRemote) {
            if (inventory.getStackInSlot(0).stackTagCompound == null)
                inventory.getStackInSlot(0).stackTagCompound = new NBTTagCompound();
            inventory.getStackInSlot(0).stackTagCompound.setInteger("coreX", getCore().xCoord);
            inventory.getStackInSlot(0).stackTagCompound.setInteger("coreY", getCore().yCoord);
            inventory.getStackInSlot(0).stackTagCompound.setInteger("coreZ", getCore().zCoord);

            inventory.setStackInSlot(inventory.getStackInSlot(0).copy(), 1);
            inventory.setStackInSlot(null, 0);
        }
    }

    @Override
    public void readFromNBT (NBTTagCompound tag) {
        super.readFromNBT(tag);
        inventory.readFromNBT(tag);
    }

    @Override
    public void writeToNBT (NBTTagCompound tag) {
        super.writeToNBT(tag);
        inventory.writeToNBT(tag);
    }

    /*******************************************************************************************************************
     ****************************************** IOpensGui Methods ******************************************************
     *******************************************************************************************************************/

    @Override
    public Object getServerGuiElement(int ID, EntityPlayer player, World world, int x, int y, int z) {
        return getCore() != null ? (player.isSneaking() && getCore().canOpen(player) ?
                new ContainerStorageRemote(player.inventory, this) :
                getCore().getServerGuiElement(ID, player, world, x, y, z)) : null;
    }

    @Override
    public Object getClientGuiElement(int ID, EntityPlayer player, World world, int x, int y, int z) {
        return getCore() != null ? (player.isSneaking() && getCore().canOpen(player) ?
                new GuiStorageRemote(new ContainerStorageRemote(player.inventory, this)) :
                getCore().getClientGuiElement(ID, player, world, x, y, z)) : null;
    }
}

File: src/main/scala/com/teambr/modularsystems/storage/tiles/TileStorageRemote.scala
package com.teambr.modularsystems.storage.tiles

import com.teambr.bookshelf.common.tiles.traits.{UpdatingTile, Inventory}
import com.teambr.modularsystems.storage.items.ItemStorageRemote
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound

/**
 * This file was created for Modular-Systems
 *
 * Modular-Systems is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Dyonovan
 * @since August 10, 2015
 */
class TileStorageRemote extends TileEntityStorageExpansion with Inventory with UpdatingTile {

    val maxDistance = 15

    override def onServerTick(): Unit = {
        if (getCore.isDefined && getStackInSlot(0) != null && getStackInSlot(1) == null) {
            val tag = getStackInSlot(0).getTagCompound
            tag.setLong("Core", getPos.toLong)
            getStackInSlot(0).setTagCompound(tag)

            setInventorySlotContents(1, getStackInSlot(0).copy)
            getStackInSlot(1).setItemDamage(getStackInSlot(0).getItemDamage)
            setInventorySlotContents(0, null)
            worldObj.markBlockForUpdate(pos)
        }
    }

    /**
     * Called after this has been added to a network
     */
    override def addedToNetwork(): Unit = { }

    /**
     * Called right before this is removed from a network
     */
    override def removedFromNetwork(): Unit = { }

    override var inventoryName: String = _

    override def hasCustomName(): Boolean = false

    override def initialSize: Int = 2

    override def isItemValidForSlot(slot: Int, stack: ItemStack): Boolean = {
        slot == 0 && stack.getItem.isInstanceOf[ItemStorageRemote]
    }

    override def readFromNBT(tag: NBTTagCompound): Unit = {
        super[Inventory].readFromNBT(tag)
        super[TileEntityStorageExpansion].readFromNBT(tag)
    }

    override def writeToNBT(tag: NBTTagCompound): Unit = {
        super[Inventory].writeToNBT(tag)
        super[TileEntityStorageExpansion].writeToNBT(tag)
    }

    override def markDirty(): Unit = {
        super[Inventory].markDirty()
        super[TileEntityStorageExpansion].markDirty()
    }
}
